autogluon/task/text_prediction/preprocessing.py

```diff
@@ -0,0 +1,556 @@
+
+        - Text fields
+            We transform text into a sequence of token_ids.
+            If there are multiple text fields, we have the following options
+            1) merge_text = True
+                We will concatenate these text fields and inserting CLS, SEP ids, i.e.
+                [CLS] text_ids1 [SEP] text_ids2 [SEP]
+            2) merge_text = False
+                We will transform each text field separately:
+                [CLS] text_ids1 [SEP], [CLS] text_ids2 [SEP], ...
+            For empty text / missing text data, we will just convert it to [CLS] [SEP]
+        - Entity fields
+            The raw entities are stored as character-level start and end offsets.
+            After the preprocessing, we will store them as the token-level
+            start + end. Different from the raw character-level start + end offsets, the
+            token-level start + end offsets will be used.
+            - token_level_start, token_level_end, span_label
+            or
+            - token_level_start, token_level_end
+        - Categorical fields
+            We transform the categorical features to its ids.
+        - Numerical fields
+            We keep the numerical features and indicate the missing value
+
+        Parameters
+        ----------
+        data
+            A single data sample.
+
+        Returns
+        -------
+        features
+            Preprocessed features. Will contain the following
+            - TEXT
+                The encoded value will be a TextTokenIdsField
+
+            - ENTITY
+                The encoded feature will be:
+                data: Shape (num_entity, 2)
+                    Each item will be (start, end)
+                if has_label:
+                    label:
+                        - Categorical: Shape (num_entity,)
+                        - Numerical: (num_entity,) + label_shape
+
+            - CATEGORICAL
+                The categorical feature. Will be an integer
+
+            - NUMERICAL
+                The numerical feature. Will be a numpy array
+        labels
+            The preprocessed labels
+        """
+        feature_fields = []
+        # Step 1: Get the text features + entity features
+        text_fields, entity_fields =\
+            process_text_entity_features(
+                data=data,
+                tokenizer=self._tokenizer,
+                text_columns=self.text_columns,
+                entity_columns=self.entity_columns,
+                max_length=self.max_length,
+                entity_props=OrderedDict([(col_name, self.column_properties[col_name])
+                                          for col_name in self.entity_columns]),
+                merge_text=self.merge_text,
+                store_token_offsets=self._store_token_offsets,
+                truncate_out_of_range=True)
+        feature_fields.extend(text_fields)
+        feature_fields.extend(entity_fields)
+
+        # Step 2: Transform all categorical columns
```

Comment:
Consider making each of these steps a dedicated class method to simplify code.

Reply:
Let me try to improve it later.